File: hw/cañarte/u3/HW29Composite&TemplateMethod/BeverageApp/src/main/java/ec/edu/espe/beverageapp/controller/CoffeeinBeverage.java
package ec.edu.espe.beverageapp.controller;

/**
 *
 * @author Saray Cañarte
 */
public abstract class CoffeeinBeverage {

    public void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) {
            addCondiments();
        }
    }

    void boilWater() {
        System.out.println("Boiling water");
    }

    void pourInCup() {
        System.out.println("Pouring into cup");
    }

    abstract void brew();

    abstract void addCondiments();

    boolean wantsCondiments() {
        return true;
    }
}

File: hw/CEDENOR/U3/HW29Composite&TemplateMethod/Python/Template/controller/CoffeeinBeverage.py
from abc import ABC, abstractmethod

class CoffeeinBeverage(ABC):
    def prepare_recipe(self):
        self.boil_water()
        self.brew()
        self.pour_in_cup()
        if self.wants_condiments():
            self.add_condiments()

    def boil_water(self):
        print("Boiling water")

    def pour_in_cup(self):
        print("Pouring into cup")

    @abstractmethod
    def brew(self):
        pass

    @abstractmethod
    def add_condiments(self):
        pass

    def wants_condiments(self):
        return True
